packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m

```diff
@@ -21,27 +21,38 @@ @interface FVPFrameUpdater : NSObject
 @property(nonatomic, weak, readonly) NSObject<FlutterTextureRegistry> *registry;
 // The output that this updater is managing.
 @property(nonatomic, weak) AVPlayerItemVideoOutput *videoOutput;
-// The last time that has been validated as avaliable according to hasNewPixelBufferForItemTime:.
-@property(nonatomic, assign) CMTime lastKnownAvailableTime;
+@property(nonatomic) CVPixelBufferRef latestPixelBuffer;
+@property(nonatomic) dispatch_queue_t pixelBufferSynchronizationQueue;
 @end
 
 @implementation FVPFrameUpdater
 - (FVPFrameUpdater *)initWithRegistry:(NSObject<FlutterTextureRegistry> *)registry {
   NSAssert(self, @"super init cannot be nil");
   if (self == nil) return nil;
   _registry = registry;
-  _lastKnownAvailableTime = kCMTimeInvalid;
   return self;
 }
 
 - (void)displayLinkFired {
   // Only report a new frame if one is actually available.
   CMTime outputItemTime = [self.videoOutput itemTimeForHostTime:CACurrentMediaTime()];
   if ([self.videoOutput hasNewPixelBufferForItemTime:outputItemTime]) {
-    _lastKnownAvailableTime = outputItemTime;
+    dispatch_async(self.pixelBufferSynchronizationQueue, ^{
+      if (self.latestPixelBuffer) {
+        CFRelease(self.latestPixelBuffer);
+      }
+      self.latestPixelBuffer = [self.videoOutput copyPixelBufferForItemTime:outputItemTime
```

Comment:
This is essentially doubling the memory usage for video output, isn't it? Why doesn't the previous approach of only storing the timestamp work? The PR description discusses the early-consume problem, but it seems like that could be addressed simply by changing `copyPixelBuffer` to prefer the last time instead of the current time.

Reply:
> This is essentially doubling the memory usage for video output, isn't it?

Can you please explain how? Maybe memory usage with the original approach would be one frame less if flutter engine deleted previous pixel buffer right before calling `copyPixelBuffer` but it would not want to do that because `copyPixelBuffer` can also return NULL and then it needs something latest to show.

Reply:
> (Also looking again, the current PR code appears to be leaking every frame it consumes.)

I cannot see it.

Reply:
Oh, I see now. The memory flow is pretty hard to follow here on the copied buffer as currently written, with the buffer sometimes freed by the frame updater, and sometimes handed off to to the engine.

Which brings us back to the initial question: why do we need to copy the buffer proactively when the display link fires instead of storing the timestamp?

Reply:
> Aren't you keeping an extra copy of the frame besides the one kept by the player and the one kept by the engine?

Both versions have a worst case number of pixel buffers 2+N where N is the number held by the player. This case is between `copyPixelBufferForItemTime` and until the engine replaces its own after `copyPixelBuffer`. Current version just can have 2+N for a little longer, especially when `displayLinkFired` is called after `copyPixelBuffer` at each frame. Btw if the player kept only a single frame then storing timestamp instead of buffer would not work.

> Which brings us back to the initial question: why do we need to copy the buffer proactively when the display link fires instead of storing the timestamp?

That would mean fetching pixel buffers from the past, especially when `displayLinkFired` is called after `copyPixelBuffer` at each frame. But this is based on undocumented and possibly wrong assumption that the player always leaves at least one past frame ready for us. What if this is not the case? Actually it is not. Seems the player periodically flushes all past pixel buffers. After there are some 3 pixel buffers in the past then the player flushes them all.

I tried an implementation which was sending timestamps instead of pixel buffers and `copyPixelBufferForItemTime` often returned NULL with timestamp for which `hasNewPixelBufferForItemTime` returned true before. It had 4x more frame drops in average during my tests compared to little modified current implementation (with `copyPixelBufferForItemTime` moved outside of `pixelBufferSynchronizationQueue`).

There are several causes of frame drops. This modified implementation minimises cases where `copyPixelBufferForItemTime` returns NULL and accidentally also frame drops caused by artefacts of using two "display links". They are caused by `displayLinkFired` and `copyPixelBuffer` changing order. Because things running on `pixelBufferSynchronizationQueue` are short (in time) and new pixel buffer is generated after some time then even when is `copyPixelBuffer` called after `displayLinkFired` at some frame (while before it was conversely) it has chance to obtain pixel buffer from `displayLinkFired` from previous frame and to not clear `_textureFrameAvailable` from this latest `displayLinkFired`. But this is of course not ideal, a more proper way would be to wait until the middle of frame and then send a new pixel buffer and call `textureFrameAvailable` but even more proper solution would be to not use two "display links".

Another cause of frame drops is when `hasNewPixelBufferForItemTime` returns false which is now prevalent. Seems this is caused by irregularities at which is called `displayLinkFired` which causes irregular timestamps returned by `CACurrentMediaTime`. I tested to use `CADisplayLink::timestamp` instead and frame drops dropped almost to zero, below 0.1%, around 20x less than current implementation (modified) on average during my tests. But this would need access to `CADisplayLink` through `FVPDisplayLink` and some other implementation for macos.

I also tested an implementation without a display link where `textureFrameAvailable` and `copyPixelBuffer` were called for each frame and it had around 3x less frame drops than the current implementation (modified). Unfortunately I could not use `CADisplayLink::timestamp` here so there were still frame drops due to irregularities. Flutter engine would need to provide something similar, or maybe it can be simply obtained by rounding `CACurrentMediaTime` down to refresh duration of display but that would then require update frequency of display from engine (or something what is doing `FVPDisplayLink` for macos).

Reply:
> That would mean fetching pixel buffers from the past, especially when `displayLinkFired` is called after `copyPixelBuffer` at each frame. But this is based on undocumented and possibly wrong assumption that the player always leaves at least one past frame ready for us. What if this is not the case? Actually it is not. Seems the player periodically flushes all past pixel buffers. After there are some 3 pixel buffers in the past then the player flushes them all.
> 
> I tried an implementation which was sending timestamps instead of pixel buffers and `copyPixelBufferForItemTime` often returned NULL with timestamp for which `hasNewPixelBufferForItemTime` returned true before. It had 4x more frame drops in average during my tests compared to little modified current implementation

Very interesting, thanks for the details! That definitely seems worth the slight memory hit.

> Another cause of frame drops is when hasNewPixelBufferForItemTime returns false which is now prevalent.

Could you file an issue with the details of what you've found here for us to follow up on in the future? It sounds like you've done a lot of great investigation here that we should be sure to capture and track.